File: 백준/Platinum/12899. 데이터 구조/데이터 구조.cc
#include <iostream>

using namespace std;

const int MAX = 2000000;
int tree[1 << 22];

void updateTree(int node, int s, int e, int idx, int v) {
  if(e < idx || idx < s) return;
  if(s == e) {
    tree[node] += v;
    return;
  }
  int mid = (s + e) >> 1;
  updateTree(node << 1, s, mid, idx, v);
  updateTree(node << 1 | 1, mid + 1, e, idx, v);
  tree[node] = tree[node << 1] + tree[node << 1 | 1];
}

int get(int node, int s, int e, int k) {
  tree[node] -= 1;
  if(s == e) return s;
  int left = tree[node << 1];
  int mid = (s + e) >> 1;
  if(k <= left) {
    return get(node << 1, s, mid, k);
  } else {
    return get(node << 1 | 1, mid + 1, e, k - left);
  }
}

void update(int val) {
  updateTree(1, 1, MAX, val, 1);
}

int query(int val) {
  return get(1, 1, MAX, val);
}

void solution() {
  int n, cmd, val;
  cin >> n;
  while(n-- > 0) {
    cin >> cmd >> val;
    if(cmd == 1) {
      update(val);
    } else {
      cout << query(val) << '\n';
    }
  }
}

int main() {
  ios::sync_with_stdio(0); cin.tie(0); cout.tie(0);
  solution();
  return 0;
}

File: 백준/Platinum/12899. 데이터 구조/데이터 구조.java
import java.io.*;
import java.util.*;

public class Main {
	static int[] tree;
	static final int MAX = 2_000_001;
	static void update(int node, int start, int end, int index) {
		if(index < start || end < index) return;
		tree[node] += 1;
		if(start == end) return;
		int mid = (start + end) / 2;
		update(2 * node, start, mid, index);
		update(2 * node + 1, mid + 1, end, index);
	}
	static int get(int node, int start, int end, int k) {
		tree[node] -= 1;
		if(start == end) return start;
		int mid = (start + end) / 2;
		if(tree[2 * node] < k) return get(2 * node + 1, mid + 1, end, k - tree[2 * node]);
		else return get(2 * node, start, mid, k);
	}
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringBuilder sb = new StringBuilder();
		StringTokenizer st = null;
		int N = Integer.parseInt(br.readLine());
		tree = new int[4 * MAX];
		for(int i = 0; i < N; i++) {
			st = new StringTokenizer(br.readLine(), " ");
			int T = Integer.parseInt(st.nextToken());
			int X = Integer.parseInt(st.nextToken());
			if(T == 1) {
				update(1, 1, MAX, X);
			} else {
				sb.append(get(1, 1, MAX, X)).append('\n');
			}
		}
		System.out.print(sb.toString());
		br.close();
	}
}